org.lflang/src/org/lflang/generator/cpp/CppActionGenerator.kt

```diff
@@ -0,0 +1,80 @@
+/*************
+ * Copyright (c) 2021, TU Dresden.
+
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ ***************/
+
+package org.lflang.generator.cpp
+
+import org.lflang.lf.Action
+import org.lflang.lf.LfPackage
+import org.lflang.lf.Reactor
+
+/** A C++ code generator for actions */
+class CppActionGenerator(private val reactor: Reactor) {
+
+    private val Action.cppType get() = if (this.isLogical) "reactor::LogicalAction" else "reactor::PhysicalAction"
+
+    private val startupName = LfPackage.Literals.TRIGGER_REF__STARTUP.name
+    private val shutdownName = LfPackage.Literals.TRIGGER_REF__SHUTDOWN.name
+
+    private fun generateDeclaration(action: Action) = "${action.cppType}<${action.targetType}> ${action.name};"
+
+    private fun generateInitializer(action: Action) =
+        if (action.isLogical) generateLogicalInitializer(action) else initializePhysicalInitializer(action)
+
+    private fun generateLogicalInitializer(action: Action): String {
+        return if (action.minSpacing != null || !action.policy.isNullOrEmpty()) {
+            TODO("How to report errors from here?")
+            //action.reportError(
+            //    "minSpacing and spacing violation policies are not yet supported for logical actions in reactor-ccp!");
+        } else {
+            val time = action.minDelay?.toTime() ?: "reactor::Duration::zero()"
+            """, ${action.name}{"${action.name}", this, $time}"""
+        }
+    }
+
+    private fun initializePhysicalInitializer(action: Action): String {
+        return if (action.minDelay != null || action.minSpacing != null || !action.policy.isNullOrEmpty()) {
```

Comment:
you could use the  expression body syntax for these methods. Also you could write them as eg
```kotlin
fun initializePhysicalInitializer(action: Action): String = with (action) {
   if (minDelay != null || minSpacing != null || ...)
    ...
}
```
this allows you to omit the `action.` prefix without making the function an extension.